### PR TITLE
[Export] First part of dynamo spec capture, spec format, capture output, check in export

### PR DESCRIPTION
**Pull request by voznesenskym**

### CODE REVIEW

File: torchdynamo/output_graph.py [hunk 1]
```diff
@@ -279,6 +281,7 @@ def compile_subgraph(self, tx, partial_convert=False):
             and self.side_effects.is_empty()
         ):
             # optimization to generate better code in a common case
+            # TODO(voz): Add spec here too before land
```
Comment (voznesenskym) [inline torchdynamo/output_graph.py:284]: Self review: This will be added before land, if the rest of the direciton looks fine

File: torchdynamo/spec.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    _append(current_collection, prior)
+                    # Now that this list is closed, we can have further additions to the prior collection
+                    # Which is the collection this collection was nested in
+                    current_collection = prior
+                else:
+                    # Closing with nothing in the stack, this is the end
+                    result = current_collection
+                    current_collection = None
+            else:
+                assert False, f"{current_element} is not yet supported"
```
Comment (voznesenskym) [inline torchdynamo/spec.py:160]: Dict supported will be added in a subsequent PR, this one has enough going on.

File: torchdynamo/codegen.py [hunk 3]
```diff
@@ -163,6 +174,13 @@ def __call__(self, value, allow_cache=True):
 
         self.top_of_stack = value
 
+    def _add_to_spec(self, value, add_open):
+        if isinstance(value, (ListVariable, TupleVariable, ConstDictVariable)):
+            if add_open:
```
Comment (jansel) [inline torchdynamo/codegen.py:179]: Should these just be two functions? (open and close)

File: torchdynamo/spec.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+from enum import Enum
+from enum import unique
+from typing import List
+
+import torch
+
+
+class Spec:
```
Comment (jansel) [inline torchdynamo/spec.py:8]: Should use full words for class names, what about `ExportSpecification`?

File: torchdynamo/spec.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        collection_stack = []
+        current_collection = None
+        idx = 0
+
+        def _append(e, collection):
+            if isinstance(collection, list):
+                collection.append(e)
+            elif isinstance(collection, tuple):
+                # TODO(voz): Optimize with forward scanning, this isnt that cheap
+                collection += (e,)
```
Comment (jansel) [inline torchdynamo/spec.py:115]: tuples are not mutable, so I think this just changes the local var and not the caller's copy.

File: torchdynamo/spec.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                    current_collection = prior
+                else:
+                    # Closing with nothing in the stack, this is the end
+                    result = current_collection
+                    current_collection = None
+            else:
+                assert False, f"{current_element} is not yet supported"
+
+            if len(reversed_elements) > 0:
+                current_element = reversed_elements.pop()
```
Comment (jansel) [inline torchdynamo/spec.py:163]: when would we expect this to happen?

File: torchdynamo/variables/lists.py [hunk 7]
```diff
@@ -132,7 +133,14 @@ def python_type(self):
 
     def reconstruct(self, codegen):
         codegen.foreach(self.items)
-        return [create_instruction("BUILD_LIST", len(self.items))]
+        list_instructions = [create_instruction("BUILD_LIST", len(self.items))]
```
Comment (jansel) [inline torchdynamo/variables/lists.py:136]: take a pass through to revert unrelated changes

File: torchdynamo/spec.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        while current_element is not None:
+            if current_element == Spec.Element.TENSOR:
+                if current_collection is None:
+                    # No current collection, make one
+                    current_collection = (tensors[idx],)
+                else:
+                    # Add tensor to current collection
+                    _append(tensors[idx], current_collection)
+                # advance tensor pointer
+                idx += 1
```
Comment (jansel) [inline torchdynamo/spec.py:130]: This seems to assume tensors are always consumed in the same order and exactly once?  Is this always true?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torchdynamo/codegen.py (hunk 3 above)
Comment (voznesenskym) [reply torchdynamo/codegen.py:179]: I had it that way, but the isinstance check felt grosser to duplicate, no strong opinions

File: torchdynamo/spec.py (hunk 5 above)
Comment (voznesenskym) [reply torchdynamo/spec.py:115]: Good catch, will add test

File: torchdynamo/variables/lists.py (hunk 7 above)
Comment (voznesenskym) [reply torchdynamo/variables/lists.py:136]: Will do

File: torchdynamo/spec.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                collection.append(e)
+            elif isinstance(collection, tuple):
+                # TODO(voz): Optimize with forward scanning, this isnt that cheap
+                collection += (e,)
+            else:
+                assert False, f"Unsupported collection type {collection.__class__}"
+
+        result = None
+        current_element = reversed_elements.pop()
+        while current_element is not None:
```
Comment (jansel) [inline torchdynamo/spec.py:121]: At least for what we have so far, this could totally just be a `pytree.TreeSpec()`.  The features supported by a pytees are a superset.

Comment (voznesenskym) [reply torchdynamo/spec.py:121]: Does pytree offer the ability to build it up iteratively, as we do with Spec? That's the biggest gap here, so far.
